### PR TITLE
Knex migration

### PR DESCRIPTION
**Pull request by vm2052**
As part of the issue #26 I changed all of the app's queries to use Knex.js
The command for installing is 
_npm install knex_

Also I changed the structure of the project, and made it layered. Data processing in one layer (services), fetching data from the DB in another (repositories). If you have any issues regarding that, please let me know.
I'm used to that kind of structure, but if you're comfortable with something else, do change it! 😄

### CODE REVIEW

**Comment by yellalena** [comment]
layered architecture is the old gold of software engineering, thank you for implementing it in our project! 😻

File: routes/users.js [hunk 1]
```diff
@@ -20,53 +22,42 @@ router.get("/users", async (req, res) => {
...
-
     if (!isValidIntegerId(id)) {
         return respondWithError(res, 400, "Invalid user id supplied");
     }
-
     try {
-        const result = await pool.query("SELECT * FROM users WHERE id = $1", [id]);
-        if (result.rows.length === 0) {
+        const userInfo = await userService.getUserById(id);
+        if (!userInfo) {
```
Comment (yellalena) [inline routes/users.js:58]: we need to get consistent with these checks. i have already seen
if result.lenth === 0
if result.rows.length === 0
if !result

File: server.js [hunk 2]
```diff
@@ -2,17 +2,18 @@
...
+const rolesService = require('./services/roles_service');
 // Middleware
 app.use(express.json()); // read documentation on what this does
 app.use(cors());
 
 (async () => { 
-  await initDb();
-  await utils.loadRolesIntoMemory(getPool());
+    await initDb();
+    console.log("1");
```
Comment (yellalena) [inline server.js:15]: :)

File: utils.js [hunk 3]
```diff
@@ -1,31 +1,5 @@
-let rolesCache = null;
 
-async function loadRolesIntoMemory(pool) {
```
Comment (yellalena) [inline utils.js:3]: I think it's more logical to keep `loadRolesIntoMemory` in utils file, since this is not something frequently used, but rather a _utility_ code called once for service optimization. can we put it back?

File: utils.js [hunk 4]
```diff
@@ -1,31 +1,5 @@
...
-    
-            console.log('Roles loaded into memory:', rolesCache);
-        }
-    } catch (error) {
-        console.error('Error loading roles:', error);
-        throw new Error('Failed to initialize roles');
-    }
-}
 
-function getRole(role_name) {
```
Comment (yellalena) [inline utils.js:23]: and this one as well, this is not related to database operations at all...

File: services/user_service.js [hunk 5]
```diff
@@ -1,15 +1,41 @@
...
+const repo = require('../repositories/user_repository');
 
-async function assignRoleToUser(userId, roleName){
-    const roleId = utils.getRole(roleName);
+async function assignRoleToUser(userId, roleName) {
     try {
-        await pool.query("INSERT INTO user_roles (user_id, role_id) VALUES ($1, $2)",
-            [userId, roleId]);
+        const roleId = rolesService.getRole(roleName);
+      await repo.assignRoleToUser(userId, roleId);
```
Comment (yellalena) [inline services/user_service.js:10]: indentation

File: services/user_service.js [hunk 6]
```diff
@@ -1,15 +1,41 @@
...
+    return await repo.getAllUsers();
+}
+async function createNewUser(name, email, about, languages) {
+    return await repo.createNewUser(name, email, about, languages);
+}
+async function getUserById(id) {
+    let userInfo = await repo.getUserInfoById(id) || {};
+
+    if (!userInfo) {
+
```
Comment (yellalena) [inline services/user_service.js:26]: please clean extra lines in this file too

File: services/user_service.js [hunk 7]
```diff
@@ -1,15 +1,41 @@
...
+        const roles = await repo.getUserRolesById(id);
+        userInfo.roles = roles.length > 0 ? roles : [];
```
Comment (yellalena) [inline services/user_service.js:28]: if roles.length is not greater than 0, isn't it already an empty array?

File: services/user_service.js [hunk 8]
```diff
@@ -1,15 +1,41 @@
...
\ No newline at end of file
+    return await repo.getAllUsers();
+}
+async function createNewUser(name, email, about, languages) {
+    return await repo.createNewUser(name, email, about, languages);
+}
+async function getUserById(id) {
+    let userInfo = await repo.getUserInfoById(id) || {};
+
+    if (!userInfo) {
```
Comment (yellalena) [inline services/user_service.js:25]: are you sure about this condition?

**Review by yellalena — CHANGES_REQUESTED** [review]

File: services/user_service.js (hunk 7 above)
Comment (marinacatbytes) [reply services/user_service.js:28]: yes looks like it, also roles.length is already checking if length of the array is > 0, so usually it's written like roles.length ? <true> : <false>

File: repositories/user_repository.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+const { getKnex } = require("../db");
+ 
+async function getAllUsers() {
+	const knex = getKnex();
+	return await knex("users").select("id", "name", "languages");
+}
+async function assignRoleToUser(user_id, role_id) {
```
[the remark this answers is not in the document]
Comment (vm2052) [reply repositories/user_repository.js:7]: i was debating between the two xD

File: repositories/user_repository.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	const knex = getKnex();
+	await knex("user_roles").insert({ role_id, user_id });
+}
+async function createNewUser(name, email, about, languages) {
+	const knex = getKnex();
+	const [user] = await knex("users").insert({ name, email, about, languages }).returning("id");
```
[the remark this answers is not in the document]
Comment (vm2052) [reply repositories/user_repository.js:13]: a single value, so just the id
without brackets i'd have to deal with an array and then extract the value

File: routes/users.js (hunk 1 above)
Comment (vm2052) [reply routes/users.js:58]: i thought result.length implies that we deal with an array of objects, no?
[ ] doesn't mean it's null like with ! 
and we're getting userInfo as just one object, not an array
I'm no JS expert I'm sorry 😭

File: db.js [hunk 11]
```diff
@@ -1,40 +1,42 @@
...
+        client: 'pg',
+        connection: {
+            host: db_host,
+            user: db_username,
+            password: db_password,
+            database: db_name,
+            port: db_port
+        },
+        connectionTimeoutMillis: 5000,
+        ssl: isLocal ? false : { rejectUnauthorized: false }
```
Comment (yellalena) [inline db.js:39]: did you try running this code? :)
access to RDS still doesn't work

please move these two lines inside the `connection` object

File: repositories/user_repository.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (yellalena) [reply repositories/user_repository.js:9]: but then on the next line you do the `user.id`🤔

**Review by yellalena — CHANGES_REQUESTED** [review]

File: repositories/user_repository.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (vm2052) [reply repositories/user_repository.js:9]: sorry i might have explained it really bad xD 
if i left it as const user = ... 
then the result of user would be [ { id:33 } ] and i'd have to parse the array and then extract the field, and then the value
with const [user] all i get is just {id:33} and to access the value i do user.id 
i'm not used to JS, i'm as confused as you are 😆 but this approach seems nicer

File: db.js (hunk 11 above)
Comment (yellalena) [reply db.js]: no it's fine, it just was't in the right place, now it works, thank you! 🙌🏻

**Review by yellalena — APPROVED** [review]
